dace/fpga_testing.py

```diff
@@ -0,0 +1,274 @@
+# Copyright 2019-2021 ETH Zurich and the DaCe authors. All rights reserved.
+from datetime import datetime
+import importlib.util
+import inspect
+import os
+import multiprocessing as mp
+from pathlib import Path
+import pytest
+import re
+import subprocess as sp
+from typing import Callable, Iterable, Optional, Tuple, Union
+
+from dace import SDFG
+from dace.config import Config, temporary_config
+
+TEST_TIMEOUT = 600  # Timeout tests after 10 minutes
+
+
+class Colors:
+    SUCCESS = "\033[92m"
+    STATUS = "\033[94m"
+    ERROR = "\033[91m"
+    BOLD = "\033[1m"
+    UNDERLINE = "\033[4m"
+    END = "\033[0m"
+
+
+def print_status(message):
+    timestamp = datetime.now().strftime("%H:%M:%S")
+    print(f"{Colors.STATUS}{Colors.BOLD}[{timestamp}]{Colors.END} {message}")
+
+
+def print_success(message):
+    timestamp = datetime.now().strftime("%H:%M:%S")
+    print(f"{Colors.SUCCESS}{Colors.BOLD}[{timestamp}]{Colors.END} {message}")
+
+
+def print_error(message):
+    timestamp = datetime.now().strftime("%H:%M:%S")
+    print(f"{Colors.ERROR}{Colors.BOLD}[{timestamp}]{Colors.END} {message}")
+
+
+def dump_logs(proc_or_logs: Union[sp.CompletedProcess, Tuple[str, str]]):
+    if isinstance(proc_or_logs, tuple):
+        log_out, log_err = proc_or_logs
+    else:
+        proc_or_logs.terminate()
+        proc_or_logs.kill()
+        try:
+            log_out, log_err = proc_or_logs.communicate(timeout=10)
+        except sp.TimeoutExpired:
+            return None  # Failed to even kill the process
+    if log_out:
+        print(log_out)
+    if log_err:
+        print(log_err)
+    return log_out, log_err
+
+
+# https://stackoverflow.com/a/33599967/2949968
```

Comment:
To propagate the exception thrown by the child process. Somehow we need to communicate that the `Process` failed, otherwise it will just exit successfully.